File: mybatis/src/test/java/data/spring/mybatis/application/required/product/ProductUseCaseTest.java
package data.spring.mybatis.application.required.product;

import data.spring.mybatis.IntegrationTestSupport;
import data.spring.mybatis.application.service.product.command.ProductSearchCommand;
import data.spring.mybatis.application.service.product.command.ProductUpdateCommand;
import data.spring.mybatis.domain.product.Product;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class ProductUseCaseTest extends IntegrationTestSupport {
    @AfterEach
    void tearDown() {
        super.productRepository.deleteAll();
    }

    @Test
    void saveProductsSuccessfully() {
        var products = List.of(
                Product.of("상품1", 20_000, 10),
                Product.of("상품2", 30_000, 20),
                Product.of("상품3", 40_000, 30)
        );
        var sut = super.productUseCase;
        sut.saveAll(products);
        var expected = List.of(
                Product.of(1L, "상품1", 20_000, 10),
                Product.of(2L, "상품2", 30_000, 20),
                Product.of(3L, "상품3", 40_000, 30)
        );

        var savedProducts = sut.findAll(ProductSearchCommand.of(null, null));

        assertThat(savedProducts).isEqualTo(expected);
    }

    @Test
    void findProductSuccessfully() {
        var product = Product.of( "리얼 마이바티스", 30_000, 100);
        var sut = super.productUseCase;
        sut.save(product);
        var expected = Product.of(1L, "리얼 마이바티스", 30_000, 100);

        var savedProduct = sut.findById(1L).get();

        assertThat(savedProduct).isEqualTo(expected);
    }

    @Test
    void findProductsSuccessfully() {
        var products = List.of(
                Product.of("리얼 마이바티스", 30_000, 100),
                Product.of("리얼 제이디비씨", 30_000, 100),
                Product.of("리얼 비쌈", 50_000, 100)
        );
        var sut = super.productUseCase;
        sut.saveAll(products);

        var savedProducts = sut.findAll(new ProductSearchCommand(null, 30_000));
        assertThat(savedProducts).hasSize(2);
        assertThat(savedProducts).extracting("productName").contains("리얼 마이바티스", "리얼 제이디비씨");
    }

    @Test
    void updateProductsSuccessfully() {
        // given
        var products = List.of(
                Product.of("상품1", 20_000, 10),
                Product.of("상품2", 30_000, 20),
                Product.of("상품3", 40_000, 30)
        );
        super.productRepository.saveAll(products);

        var updateCommands = List.of(
                ProductUpdateCommand.of(1L, "상품4", null, null),
                ProductUpdateCommand.of(2L, "상품5", null, null),
                ProductUpdateCommand.of(3L, "상품6", null, null)
        );
        var sut = super.productUseCase;

        // when
        sut.updateAll(updateCommands);

        // then
        var savedProducts = super.productRepository.findAll(ProductSearchCommand.of(null, null));
        assertThat(savedProducts).extracting("productName").containsExactly("상품4", "상품5", "상품6");
    }
}

File: mybatis/src/test/kotlin/data/spring/mybatis/application/provided/product/ProductUseCaseTest.kt
package data.spring.mybatis.application.provided.product

import data.spring.mybatis.IntegrationTestSupport
import data.spring.mybatis.application.exception.NoDataFoundException
import data.spring.mybatis.application.service.product.command.ProductSearchCommand
import data.spring.mybatis.application.service.product.command.ProductUpdateCommand
import data.spring.mybatis.domain.product.Product
import data.spring.mybatis.domain.testClock
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

class ProductUseCaseTest: IntegrationTestSupport() {
    val now = LocalDateTime.now(testClock())

    @AfterEach
    fun tearDown() {
        super.productUseCase.deleteAll()
    }

    @Test
    fun saveProductsSuccessfully() {
        val sut = super.productUseCase
        val products = listOf(
            Product(productName = "상품1", price = 20000, quantity = 10, createdAt = now, updatedAt = now),
            Product(productName = "상품2", price = 30000, quantity = 20, createdAt = now, updatedAt = now),
            Product(productName = "상품3", price = 40000, quantity = 30, createdAt = now, updatedAt = now)
        )
        sut.saveAll(products)
        val expected = listOf(
            Product(1L, "상품1", 20000, 10, createdAt = now, updatedAt = now),
            Product(2L, "상품2", 30000, 20, createdAt = now, updatedAt = now),
            Product(3L, "상품3", 40000, 30, createdAt = now, updatedAt = now)
        )

        val savedProducts = sut.findWithCond(ProductSearchCommand(null, null))

        assertThat(savedProducts).isEqualTo(expected)
    }

    @Test
    fun findProductSuccessfully() {
        val sut = super.productUseCase
        val product = Product(productName = "리얼 마이바티스", price = 30000, quantity = 100, createdAt = now, updatedAt = now)
        sut.save(product)
        val expected = Product(productId = 1L, productName = "리얼 마이바티스", price = 30000, quantity = 100,
            createdAt = now, updatedAt = now)

        val savedProduct = sut.findById(1L)

        assertThat(savedProduct).isEqualTo(expected)
    }

    @Test
    fun findProductsSuccessfully() {
        val sut = super.productUseCase
        val products = listOf(
            Product(productName = "리얼 마이바티스", price = 30000, quantity = 100),
            Product(productName = "리얼 제이디비씨", price = 30000, quantity = 100),
            Product(productName = "리얼 비쌈", price = 50000, quantity = 100)
        )
        sut.saveAll(products)

        val savedProducts = sut.findWithCond(ProductSearchCommand(productName = null, maxPrice = 30000))
        assertThat(savedProducts).hasSize(2)
        assertThat(savedProducts).extracting("productName").contains("리얼 마이바티스", "리얼 제이디비씨")
    }

    @Test
    fun updateProductsSuccessfully() {
        // given
        val sut = super.productUseCase
        val products = listOf(
            Product(productName = "상품1", price = 20000, quantity = 10),
            Product(productName = "상품2", price = 30000, quantity = 20),
            Product(productName = "상품3", price = 40000, quantity = 30)
        )
        sut.saveAll(products)
        val updateCommands = listOf(
            ProductUpdateCommand(1L, "상품4", null, null),
            ProductUpdateCommand(2L, "상품5", null, null),
            ProductUpdateCommand(3L, "상품6", null, null)
        )

        // when
        sut.updateList(updateCommands)

        // then
        val savedProducts = sut.findWithCond(ProductSearchCommand(productName = null, maxPrice = null))
        assertThat(savedProducts).extracting("productName").containsExactly("상품4", "상품5", "상품6")
    }

    @Test
    fun findWhenNoProduct() {
        val sut = super.productUseCase

        assertThatThrownBy { sut.findById(1L) }
                .isInstanceOf(NoDataFoundException::class.java);
    }

    @Test
    fun findAllWhenNoProduct() {
        val sut = super.productUseCase

        assertThatThrownBy { sut.findWithCond(ProductSearchCommand(productName = null, maxPrice = null)) }
            .isInstanceOf(NoDataFoundException::class.java)
    }
}